File: clients/python/uiform/resources/schemas.py
from io import IOBase
from pathlib import Path
from typing import Any, List, Optional, Sequence

import PIL.Image
from pydantic import BaseModel

from .._resource import AsyncAPIResource, SyncAPIResource
from .._utils.ai_models import assert_valid_model_schema_generation
from .._utils.json_schema import load_json_schema
from .._utils.mime import prepare_mime_document_list
from ..types.modalities import Modality
from ..types.mime import MIMEData
from ..types.schemas.generate import GenerateSchemaRequest
from ..types.schemas.enhance import EnhanceSchemaConfig, EnhanceSchemaConfigDict, EnhanceSchemaRequest
from ..types.schemas.evaluate import EvaluateSchemaRequest, EvaluateSchemaResponse
from ..types.schemas.object import Schema
from ..types.standards import PreparedRequest
from openai.types.chat.chat_completion_reasoning_effort import ChatCompletionReasoningEffort


class SchemasMixin:
    def prepare_generate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        instructions: str | None = None,
        model: str = "gpt-4o-2024-11-20",
        temperature: float = 0,
        modality: Modality = "native",
    ) -> PreparedRequest:
        assert_valid_model_schema_generation(model)
        mime_documents = prepare_mime_document_list(documents)
        data = {
            "documents": [doc.model_dump() for doc in mime_documents],
            "instructions": instructions if instructions else None,
            "model": model,
            "temperature": temperature,
            "modality": modality,
        }
        GenerateSchemaRequest.model_validate(data)
        return PreparedRequest(method="POST", url="/v1/schemas/generate", data=data)

    def prepare_evaluate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        json_schema: dict[str, Any],
        ground_truths: list[dict[str, Any]] | None = None,
        model: str = "gpt-4o-mini",
        temperature: float = 0.0,
        reasoning_effort: ChatCompletionReasoningEffort = "medium",
        modality: Modality = "native",
        image_resolution_dpi: int = 96,
        browser_canvas: str = "A4",
        n_consensus: int = 1,
    ) -> PreparedRequest:
        # Assert that if ground_truths is not None, it has the same length as documents
        if ground_truths is not None and len(documents) != len(ground_truths):
            raise ValueError("Number of documents must match number of ground truths")

        mime_documents = prepare_mime_document_list(documents)
        data = {
            "documents": [doc.model_dump() for doc in mime_documents],
            "ground_truths": ground_truths,
            "model": model,
            "temperature": temperature,
            "reasoning_effort": reasoning_effort,
            "modality": modality,
            "image_resolution_dpi": image_resolution_dpi,
            "browser_canvas": browser_canvas,
            "n_consensus": n_consensus,
            "json_schema": json_schema,
        }
        EvaluateSchemaRequest.model_validate(data)
        return PreparedRequest(method="POST", url="/v1/schemas/evaluate", data=data)

    def prepare_enhance(
        self,
        json_schema: dict[str, Any] | Path | str,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        ground_truths: list[dict[str, Any]] | None,
        instructions: str | None,
        model: str,
        temperature: float,
        modality: Modality,
        flat_likelihoods: list[dict[str, float]] | dict[str, float] | None,
        tools_config: EnhanceSchemaConfig,
    ) -> PreparedRequest:
        assert_valid_model_schema_generation(model)
        mime_documents = prepare_mime_document_list(documents)
        loaded_json_schema = load_json_schema(json_schema)
        data = {
            "json_schema": loaded_json_schema,
            "documents": [doc.model_dump() for doc in mime_documents],
            "ground_truths": ground_truths,
            "instructions": instructions if instructions else None,
            "model": model,
            "temperature": temperature,
            "modality": modality,
            "flat_likelihoods": flat_likelihoods,
            "tools_config": tools_config.model_dump(),
        }
        EnhanceSchemaRequest.model_validate(data)
        return PreparedRequest(method="POST", url="/v1/schemas/enhance", data=data)

    def prepare_get(self, schema_id: str) -> PreparedRequest:
        return PreparedRequest(method="GET", url=f"/v1/schemas/{schema_id}")


class Schemas(SyncAPIResource, SchemasMixin):
    def load(self, json_schema: dict[str, Any] | Path | str | None = None, pydantic_model: type[BaseModel] | None = None) -> Schema:
        """Load a schema from a JSON schema.

        Args:
            json_schema: The JSON schema to load
        """
        if json_schema:
            return Schema(json_schema=load_json_schema(json_schema))
        elif pydantic_model:
            return Schema(pydantic_model=pydantic_model)
        else:
            raise ValueError("Either json_schema or pydantic_model must be provided")

    def generate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        instructions: str | None = None,
        model: str = "gpt-4o-2024-11-20",
        temperature: float = 0,
        modality: Modality = "native",
    ) -> Schema:
        """
        Generate a complete JSON schema by analyzing the provided documents.

        The generated schema includes X-Prompts for enhanced LLM interactions:
        - X-SystemPrompt: Defines high-level instructions and context for consistent LLM behavior
        - X-ReasoningPrompt: Creates auxiliary reasoning fields for complex data processing

        Args:
            documents: List of documents (as MIMEData) to analyze

        Returns:
            dict[str, Any]: Generated JSON schema with X-Prompts based on document analysis

        Raises:
            HTTPException if the request fails
        """

        prepared_request = self.prepare_generate(documents, instructions, model, temperature, modality)
        response = self._client._prepared_request(prepared_request)
        return Schema.model_validate(response)

    def evaluate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        json_schema: dict[str, Any],
        ground_truths: list[dict[str, Any]] | None = None,
        model: str = "gpt-4o-mini",
        temperature: float = 0.0,
        reasoning_effort: ChatCompletionReasoningEffort = "medium",
        modality: Modality = "native",
        image_resolution_dpi: int = 96,
        browser_canvas: str = "A4",
        n_consensus: int = 1,
    ) -> EvaluateSchemaResponse:
        """
        Evaluate a schema by performing extractions on provided documents.
        If ground truths are provided, compares extractions against them.
        Otherwise, uses consensus likelihoods as metrics.

        Args:
            documents: List of documents to evaluate against
            json_schema: The JSON schema to evaluate
            ground_truths: Optional list of ground truth dictionaries to compare against
            model: The model to use for extraction
            temperature: The temperature to use for extraction
            reasoning_effort: The reasoning effort to use for extraction
            modality: The modality to use for extraction
            image_resolution_dpi: The DPI of the image. Defaults to 96.
            browser_canvas: The canvas size of the browser. Must be one of:
                - "A3" (11.7in x 16.54in)
                - "A4" (8.27in x 11.7in) 
                - "A5" (5.83in x 8.27in)
                Defaults to "A4".
            n_consensus: Number of consensus rounds to perform

        Returns:
            EvaluateSchemaResponse: Dictionary containing evaluation metrics for each document.
            Each metric includes:
            - similarity: Average similarity/likelihood score
            - similarities: Per-field similarity/likelihood scores
            - flat_similarities: Flattened per-field similarity/likelihood scores
            - aligned_* versions of the above metrics

        Raises:
            ValueError: If ground_truths is provided and its length doesn't match documents
            HTTPException: If the request fails or if there are too many documents
        """
        prepared_request = self.prepare_evaluate(
            documents=documents,
            json_schema=json_schema,
            ground_truths=ground_truths,
            model=model,
            temperature=temperature,
            reasoning_effort=reasoning_effort,
            modality=modality,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            n_consensus=n_consensus,
        )
        response = self._client._prepared_request(prepared_request)
        return EvaluateSchemaResponse.model_validate(response)

    def enhance(
        self,
        json_schema: dict[str, Any] | Path | str,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        ground_truths: list[dict[str, Any]] | None = None,
        instructions: str | None = None,
        model: str = "gpt-4o-2024-11-20",
        temperature: float = 0,
        modality: Modality = "native",
        flat_likelihoods: list[dict[str, float]] | dict[str, float] | None = None,
        tools_config: EnhanceSchemaConfigDict | None = None,
    ) -> Schema:
        prepared_request = self.prepare_enhance(
            json_schema, documents, ground_truths, instructions, model, temperature, modality, flat_likelihoods, EnhanceSchemaConfig.model_validate(tools_config or {})
        )
        response = self._client._prepared_request(prepared_request)
        return Schema(json_schema=response["json_schema"])


class AsyncSchemas(AsyncAPIResource, SchemasMixin):
    async def load(self, json_schema: dict[str, Any] | Path | str | None = None, pydantic_model: type[BaseModel] | None = None) -> Schema:
        """Load a schema from a JSON schema.

        Args:
            json_schema: The JSON schema to load
            pydantic_model: The Pydantic model to load
        """
        if json_schema:
            return Schema(json_schema=load_json_schema(json_schema))
        elif pydantic_model:
            return Schema(pydantic_model=pydantic_model)
        else:
            raise ValueError("Either json_schema or pydantic_model must be provided")

    """Schemas Asyncronous API wrapper"""

    async def get(self, schema_id: str) -> Schema:
        """Retrieve a schema by ID.

        Args:
            schema_id: The ID of the schema to retrieve

        Returns:
            Schema: The retrieved schema object
        """

        prepared_request = self.prepare_get(schema_id)
        response = await self._client._prepared_request(prepared_request)
        return Schema.model_validate(response)

    async def generate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        instructions: str | None = None,
        model: str = "gpt-4o-2024-11-20",
        temperature: float = 0.0,
        modality: Modality = "native",
    ) -> Schema:
        """
        Generate a complete JSON schema by analyzing the provided documents.

        The generated schema includes X-Prompts for enhanced LLM interactions:
        - X-SystemPrompt: Defines high-level instructions and context for consistent LLM behavior
        - X-FieldPrompt: Enhances standard description fields with specific extraction guidance
        - X-ReasoningPrompt: Creates auxiliary reasoning fields for complex data processing

        Args:
            documents: List of documents (as MIMEData) to analyze

        Returns:
            dict[str, Any]: Generated JSON schema with X-Prompts based on document analysis

        Raises
            HTTPException if the request fails
        """
        prepared_request = self.prepare_generate(documents, instructions, model, temperature, modality)
        response = await self._client._prepared_request(prepared_request)
        return Schema.model_validate(response)

    async def evaluate(
        self,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        json_schema: dict[str, Any],
        ground_truths: list[dict[str, Any]] | None = None,
        model: str = "gpt-4o-mini",
        temperature: float = 0.0,
        reasoning_effort: ChatCompletionReasoningEffort = "medium",
        modality: Modality = "native",
        image_resolution_dpi: int = 96,
        browser_canvas: str = "A4",
        n_consensus: int = 1,
    ) -> EvaluateSchemaResponse:
        """
        Evaluate a schema by performing extractions on provided documents.
        If ground truths are provided, compares extractions against them.
        Otherwise, uses consensus likelihoods as metrics.

        Args:
            documents: List of documents to evaluate against
            json_schema: The JSON schema to evaluate
            ground_truths: Optional list of ground truth dictionaries to compare against
            model: The model to use for extraction
            temperature: The temperature to use for extraction
            reasoning_effort: The reasoning effort to use for extraction
            modality: The modality to use for extraction
            image_resolution_dpi: The DPI of the image. Defaults to 96.
            browser_canvas: The canvas size of the browser. Must be one of:
                - "A3" (11.7in x 16.54in)
                - "A4" (8.27in x 11.7in) 
                - "A5" (5.83in x 8.27in)
                Defaults to "A4".
            n_consensus: Number of consensus rounds to perform

        Returns:
            EvaluateSchemaResponse: Dictionary containing evaluation metrics for each document.
            Each metric includes:
            - similarity: Average similarity/likelihood score
            - similarities: Per-field similarity/likelihood scores
            - flat_similarities: Flattened per-field similarity/likelihood scores
            - aligned_* versions of the above metrics

        Raises:
            ValueError: If ground_truths is provided and its length doesn't match documents
            HTTPException: If the request fails or if there are too many documents
        """
        prepared_request = self.prepare_evaluate(
            documents=documents,
            json_schema=json_schema,
            ground_truths=ground_truths,
            model=model,
            temperature=temperature,
            reasoning_effort=reasoning_effort,
            modality=modality,
            image_resolution_dpi=image_resolution_dpi,
            browser_canvas=browser_canvas,
            n_consensus=n_consensus,
        )
        response = await self._client._prepared_request(prepared_request)
        return EvaluateSchemaResponse.model_validate(response)

    async def enhance(
        self,
        json_schema: dict[str, Any] | Path | str,
        documents: Sequence[Path | str | bytes | MIMEData | IOBase | PIL.Image.Image],
        ground_truths: list[dict[str, Any]] | None = None,
        instructions: str | None = None,
        model: str = "gpt-4o-2024-11-20",
        temperature: float = 0,
        modality: Modality = "native",
        flat_likelihoods: list[dict[str, float]] | dict[str, float] | None = None,
        tools_config: EnhanceSchemaConfigDict | None = None,
    ) -> Schema:
        prepared_request = self.prepare_enhance(
            json_schema, documents, ground_truths, instructions, model, temperature, modality, flat_likelihoods, EnhanceSchemaConfig.model_validate(tools_config or {})
        )
        response = await self._client._prepared_request(prepared_request)
        return Schema(json_schema=response["json_schema"])

File: clients/node/src/resources/schemas.ts
import fs from 'fs';
import path from 'path';
import { SyncAPIResource, AsyncAPIResource } from '../resource.js';
import { PreparedRequest } from '../types/standards.js';
import { Schema } from '../types/schemas/object.js';
import { loadJsonSchema } from '../utils/json_schema_utils.js';

// TypeScript equivalents of Python types
export type MIMEData = string | Buffer | { filename: string; content: Buffer; mimeType: string };
export type Modality = 'native' | 'text';
export type BrowserCanvas = 'A3' | 'A4' | 'A5';
export type ChatCompletionReasoningEffort = 'low' | 'medium' | 'high';

export interface GenerateSchemaRequest {
  documents: MIMEData[];
  instructions?: string | null;
  model: string;
  temperature: number;
  modality: Modality;
  reasoning_effort: ChatCompletionReasoningEffort;
}

export interface EvaluateSchemaRequest {
  documents: MIMEData[];
  json_schema: Record<string, any>;
  ground_truths?: Array<Record<string, any>> | null;
  model: string;
  reasoning_effort: ChatCompletionReasoningEffort;
  modality: Modality;
  image_resolution_dpi: number;
  browser_canvas: BrowserCanvas;
  n_consensus: number;
}

export interface EvaluateSchemaResponse {
  [key: string]: any; // Evaluation metrics
}

export interface EnhanceSchemaConfig {
  [key: string]: any;
}

export interface EnhanceSchemaRequest {
  json_schema: Record<string, any>;
  documents: MIMEData[];
  ground_truths?: Array<Record<string, any>> | null;
  instructions?: string | null;
  model: string;
  temperature: number;
  modality: Modality;
  flat_likelihoods?: Array<Record<string, number>> | Record<string, number> | null;
  tools_config: EnhanceSchemaConfig;
}

// Helper functions (simplified versions of Python utilities)
function assertValidModelSchemaGeneration(model: string): void {
  // Validate model for schema generation
  const validModels = ['gpt-4o-2024-11-20', 'gpt-4o-mini', 'gpt-4o'];
  if (!validModels.includes(model)) {
    throw new Error(`Model ${model} not valid for schema generation`);
  }
}

function prepareMimeDocumentList(documents: Array<string | Buffer | MIMEData>): MIMEData[] {
  return documents.map(doc => {
    if (typeof doc === 'string') {
      // If it's a file path, read the file
      if (fs.existsSync(doc)) {
        return {
          filename: path.basename(doc),
          content: fs.readFileSync(doc),
          mimeType: getMimeType(doc),
        };
      }
      // Otherwise treat as text content
      return doc;
    }
    return doc as MIMEData;
  });
}

function getMimeType(filePath: string): string {
  const ext = path.extname(filePath).toLowerCase();
  const mimeTypes: Record<string, string> = {
    '.pdf': 'application/pdf',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png',
    '.txt': 'text/plain',
    '.json': 'application/json',
  };
  return mimeTypes[ext] || 'application/octet-stream';
}


class SchemasMixin {
  public prepareGenerate(
    documents: Array<string | Buffer | MIMEData>,
    instructions?: string | null,
    model: string = 'gpt-4o-2024-11-20',
    temperature: number = 0,
    modality: Modality = 'native',
    reasoning_effort: ChatCompletionReasoningEffort = 'medium'
  ): PreparedRequest {
    assertValidModelSchemaGeneration(model);
    const mimeDocuments = prepareMimeDocumentList(documents);
    const request: GenerateSchemaRequest = {
      documents: mimeDocuments,
      instructions: instructions || undefined,
      model,
      temperature,
      modality,
      reasoning_effort,
    };
    return {
      method: 'POST',
      url: '/v1/schemas/generate',
      data: request,
    };
  }

  public prepareEvaluate(
    documents: Array<string | Buffer | MIMEData>,
    jsonSchema: Record<string, any>,
    groundTruths?: Array<Record<string, any>> | null,
    model: string = 'gpt-4o-mini',
    reasoning_effort: ChatCompletionReasoningEffort = 'medium',
    modality: Modality = 'native',
    imageResolutionDpi: number = 96,
    browserCanvas: BrowserCanvas = 'A4',
    nConsensus: number = 1
  ): PreparedRequest {
    const mimeDocuments = prepareMimeDocumentList(documents);
    const request: EvaluateSchemaRequest = {
      documents: mimeDocuments,
      json_schema: jsonSchema,
      ground_truths: groundTruths || undefined,
      model,
      reasoning_effort,
      modality,
      image_resolution_dpi: imageResolutionDpi,
      browser_canvas: browserCanvas,
      n_consensus: nConsensus,
    };
    return {
      method: 'POST',
      url: '/v1/schemas/evaluate',
      data: request,
    };
  }

  public prepareEnhance(
    jsonSchema: Record<string, any> | string,
    documents: Array<string | Buffer | MIMEData>,
    groundTruths?: Array<Record<string, any>> | null,
    instructions?: string | null,
    model: string = 'gpt-4o-mini',
    temperature: number = 0.0,
    modality: Modality = 'native',
    flatLikelihoods?: Array<Record<string, number>> | Record<string, number> | null,
    toolsConfig: EnhanceSchemaConfig = {}
  ): PreparedRequest {
    assertValidModelSchemaGeneration(model);
    const mimeDocuments = prepareMimeDocumentList(documents);
    const loadedJsonSchema = loadJsonSchema(jsonSchema);
    const request: EnhanceSchemaRequest = {
      json_schema: loadedJsonSchema,
      documents: mimeDocuments,
      ground_truths: groundTruths || undefined,
      instructions: instructions || undefined,
      model,
      temperature,
      modality,
      flat_likelihoods: flatLikelihoods || undefined,
      tools_config: toolsConfig,
    };
    return {
      method: 'POST',
      url: '/v1/schemas/enhance',
      data: request,
    };
  }

  public prepareGet(schemaId: string): PreparedRequest {
    return {
      method: 'GET',
      url: `/v1/schemas/${schemaId}`,
    };
  }
}

export class Schemas extends SyncAPIResource {
  private mixin = new SchemasMixin();

  load(jsonSchema?: Record<string, any> | string | null, pydanticModel?: any | null): Schema {
    if (jsonSchema) {
      return new Schema({ json_schema: loadJsonSchema(jsonSchema) });
    } else if (pydanticModel) {
      return new Schema({ pydanticModel });
    } else {
      throw new Error('Either json_schema or pydantic_model must be provided');
    }
  }

  async generate(
    documents: Array<string | Buffer | MIMEData>,
    instructions?: string | null,
    model: string = 'gpt-4o-2024-11-20',
    temperature: number = 0,
    modality: Modality = 'native'
  ): Promise<Schema> {
    const preparedRequest = this.mixin.prepareGenerate(documents, instructions, model, temperature, modality);
    const response = await this._client._preparedRequest(preparedRequest);
    return Schema.validate(response);
  }

  async evaluate(
    documents: Array<string | Buffer | MIMEData>,
    jsonSchema: Record<string, any>,
    groundTruths?: Array<Record<string, any>> | null,
    model: string = 'gpt-4o-mini',
    reasoning_effort: ChatCompletionReasoningEffort = 'medium',
    modality: Modality = 'native',
    imageResolutionDpi: number = 96,
    browserCanvas: BrowserCanvas = 'A4',
    nConsensus: number = 1
  ): Promise<EvaluateSchemaResponse> {
    const preparedRequest = this.mixin.prepareEvaluate(
      documents,
      jsonSchema,
      groundTruths,
      model,
      reasoning_effort,
      modality,
      imageResolutionDpi,
      browserCanvas,
      nConsensus
    );
    const response = await this._client._preparedRequest(preparedRequest);
    return response;
  }

  async enhance(
    jsonSchema: Record<string, any> | string,
    documents: Array<string | Buffer | MIMEData>,
    groundTruths?: Array<Record<string, any>> | null,
    instructions?: string | null,
    model: string = 'gpt-4o-2024-11-20',
    temperature: number = 0,
    modality: Modality = 'native',
    flatLikelihoods?: Array<Record<string, number>> | Record<string, number> | null,
    toolsConfig?: EnhanceSchemaConfig
  ): Promise<Schema> {
    const _toolsConfig = toolsConfig || {};
    const preparedRequest = this.mixin.prepareEnhance(
      jsonSchema,
      documents,
      groundTruths,
      instructions,
      model,
      temperature,
      modality,
      flatLikelihoods,
      _toolsConfig
    );
    const response = await this._client._preparedRequest(preparedRequest);
    return new Schema({ json_schema: response.json_schema });
  }
}

export class AsyncSchemas extends AsyncAPIResource {
  private mixin = new SchemasMixin();

  load(jsonSchema?: Record<string, any> | string | null, pydanticModel?: any | null): Schema {
    if (jsonSchema) {
      return new Schema({ json_schema: loadJsonSchema(jsonSchema) });
    } else if (pydanticModel) {
      return new Schema({ pydanticModel });
    } else {
      throw new Error('Either json_schema or pydantic_model must be provided');
    }
  }

  async get(schemaId: string): Promise<Schema> {
    const preparedRequest = this.mixin.prepareGet(schemaId);
    const response = await this._client._preparedRequest(preparedRequest);
    return Schema.validate(response);
  }

  async generate(
    documents: Array<string | Buffer | MIMEData>,
    instructions?: string | null,
    model: string = 'gpt-4o-2024-11-20',
    temperature: number = 0.0,
    modality: Modality = 'native'
  ): Promise<Schema> {
    const preparedRequest = this.mixin.prepareGenerate(
      documents,
      instructions,
      model,
      temperature,
      modality
    );
    const response = await this._client._preparedRequest(preparedRequest);
    return Schema.validate(response);
  }

  async evaluate(
    documents: Array<string | Buffer | MIMEData>,
    jsonSchema: Record<string, any>,
    groundTruths?: Array<Record<string, any>> | null,
    model: string = 'gpt-4o-mini',
    reasoning_effort: ChatCompletionReasoningEffort = 'medium',
    modality: Modality = 'native',
    imageResolutionDpi: number = 96,
    browserCanvas: BrowserCanvas = 'A4',
    nConsensus: number = 1
  ): Promise<EvaluateSchemaResponse> {
    const preparedRequest = this.mixin.prepareEvaluate(
      documents,
      jsonSchema,
      groundTruths,
      model,
      reasoning_effort,
      modality,
      imageResolutionDpi,
      browserCanvas,
      nConsensus
    );
    const response = await this._client._preparedRequest(preparedRequest);
    return response;
  }

  async enhance(
    jsonSchema: Record<string, any> | string,
    documents: Array<string | Buffer | MIMEData>,
    groundTruths?: Array<Record<string, any>> | null,
    instructions?: string | null,
    model: string = 'gpt-4o-2024-11-20',
    temperature: number = 0,
    modality: Modality = 'native',
    flatLikelihoods?: Array<Record<string, number>> | Record<string, number> | null,
    toolsConfig?: EnhanceSchemaConfig
  ): Promise<Schema> {
    const _toolsConfig = toolsConfig || {};
    const preparedRequest = this.mixin.prepareEnhance(
      jsonSchema,
      documents,
      groundTruths,
      instructions,
      model,
      temperature,
      modality,
      flatLikelihoods,
      _toolsConfig
    );
    const response = await this._client._preparedRequest(preparedRequest);
    return new Schema({ json_schema: response.json_schema });
  }
}